services/121-service/src/transaction-job-processors/transaction-job-processors.service.ts

```diff
@@ -224,18 +228,119 @@ export class TransactionJobProcessorsService {
     });
   }
 
+  public async processSafaricomTransactionJob(
+    input: SafaricomTransactionJobDto,
+  ): Promise<void> {
+    // TODO: update/remove the numbered steps below, which were initially written down as a general structure based on Intersolve
+
+    // 1. Get registration details needed
+    // TODO: this is duplicate code with Visa-method > simplify
+    const registration =
+      await this.registrationScopedRepository.getByReferenceId({
+        referenceId: input.referenceId,
+      });
+    if (!registration) {
+      throw new Error(
+        `Registration was not found for referenceId ${input.referenceId}`,
+      );
+    }
+    const oldRegistration = structuredClone(registration);
+    const financialServiceProvider =
+      await this.financialServiceProviderRepository.getByName(
+        FinancialServiceProviderName.safaricom,
+      );
+    if (!financialServiceProvider) {
+      throw new Error('Financial Service Provider not found');
+    }
+
+    // 2. Check if all required properties are present. If not, create a failed transaction and throw an error.
+    for (const [name, value] of Object.entries(input)) {
+      // TODO: make some properties optional like in Visa, but why?
+
+      // Define "empty" based on your needs. Here, we check for null, undefined, or an empty string.
+      if (value === null || value === undefined || value === '') {
+        const errorText = `Property ${name} is undefined`;
+        await this.createTransactionAndUpdateRegistration({
+          programId: input.programId,
+          paymentNumber: input.paymentNumber,
+          userId: input.userId,
+          calculatedTransferAmountInMajorUnit: input.transactionAmount,
+          financialServiceProviderId: financialServiceProvider.id,
+          registration,
+          oldRegistration,
+          isRetry: input.isRetry,
+          status: StatusEnum.error,
+          errorText,
+        });
+        return;
+      }
+    }
+
+    // 3. Start the transfer, save error transaction on failure
+    let safaricomDoTransferResult: DoTransferReturnParams;
+    try {
+      safaricomDoTransferResult = await this.safaricomService.doTransfer({
```

Comment:
On first try (so not retry): first create the SafaricomTransfer Entity and the Transaction Entity (and await), and then call doTransfer(). So that the callback always has data in the database). See Sequence Diagram in Miro.

Reply:
see separate task